### PR TITLE
added react-native binding

### PR DESCRIPTION
**Pull request by ksyeo1010**
- updated react-native binding
- updated react-native demos

### CODE REVIEW

File: binding/react-native/android/src/main/java/ai/picovoice/reactnative/porcupine/PorcupineModule.java [hunk 1]
```diff
@@ -132,7 +140,7 @@ public void process(String handle, ReadableArray pcmArray, Promise promise) {
       int result = porcupine.process(buffer);
       promise.resolve(result);
     } catch (PorcupineException e) {
-      promise.reject(e.toString());
+      promise.reject(e);
```
Comment (laves) [inline binding/react-native/android/src/main/java/ai/picovoice/reactnative/porcupine/PorcupineModule.java:143]: again, there must be a conversion we can do to bring the right exception to the RN layer

File: binding/react-native/android/src/main/java/ai/picovoice/reactnative/porcupine/PorcupineModule.java [hunk 2]
```diff
@@ -119,7 +127,7 @@ public void process(String handle, ReadableArray pcmArray, Promise promise) {
     try {
 
       if (!porcupinePool.containsKey(handle)) {
-        promise.reject("Invalid Porcupine handle provided to native module.");
+        promise.reject(new Exception("Invalid Porcupine handle provided to native module."));
```
Comment (laves) [inline binding/react-native/android/src/main/java/ai/picovoice/reactnative/porcupine/PorcupineModule.java:130]: should probably be some form of porcupine exception

File: binding/react-native/ios/Porcupine.swift [hunk 3]
```diff
@@ -10,6 +10,12 @@
 //
 
 import PvPorcupine
+import Foundation
+
+enum PvPorcupineError: Error {
+    case RuntimeError(_ message: String)
```
Comment (laves) [inline binding/react-native/ios/Porcupine.swift:16]: same comment as on Android. We should try to find a way to maintain custom exceptions all the way up.

File: binding/react-native/ios/Porcupine.swift [hunk 4]
```diff
@@ -86,4 +120,18 @@ class PvPorcupine: NSObject {
             reject("PvPorcupine:process", "Invalid Porcupine handle provided to native module.", nil)
         }
     }
+
+    private func extractResource(_ filePath: String) throws -> String {
```
Comment (laves) [inline binding/react-native/ios/Porcupine.swift:124]: extractResource isn't really an accurate name anymore. Probably should call it getResourcePath or something

File: binding/react-native/src/porcupinemanager.tsx [hunk 5]
```diff
@@ -14,11 +14,13 @@ import { EventSubscription, NativeEventEmitter } from 'react-native';
 
 import Porcupine from './porcupine';
 export type DetectionCallback = (keywordIndex: number) => void;
+export type ProcessErrorCallback = (error: unknown) => void;
```
Comment (laves) [inline binding/react-native/src/porcupinemanager.tsx:17]: can we type this error as a Porcupine error of some sort?

File: binding/react-native/src/porcupine.tsx [hunk 6]
```diff
@@ -32,11 +34,13 @@ class Porcupine {
    * @returns An instance of the engine.
    */
   public static async fromKeywords(
+    accessKey: string,
```
Comment (laves) [inline binding/react-native/src/porcupine.tsx:37]: I think to bring this in line with our other SDKs, we should use an enum for built-in keywords and call this function `fromBuiltInKeywords`. Also, "default keywords" -> "built-in keywords" and when you're providing a path to a file we're going to call it a "custom keyword"

File: binding/react-native/src/porcupine.tsx [hunk 7]
```diff
@@ -80,18 +88,25 @@ class Porcupine {
    * @returns An instance of the engine.
    */
   private static async create(
+    accessKey: string,
```
Comment (laves) [inline binding/react-native/src/porcupine.tsx:91]: once you've changed keywords to a BuiltInKeyword type. This private create constructor can just take keyword paths, with `fromBuiltInKeywords` doing the conversion from BuiltInKeyword to string path.

File: demo/react-native/App.tsx [hunk 8]
```diff
@@ -23,12 +23,15 @@ type State = {
...
   backgroundColour: string;
+  isError: boolean;
+  errorMessage: string;
 };
 
 export default class App extends Component<Props, State> {
   _porcupineManager: PorcupineManager | undefined;
   _detectionColour: string = '#00E5C3';
   _defaultColour: string = '#F5FCFF';
+  _accessKey: string = "YOUR_ACCESS_KEY_HERE" // AccessKey obtained from Picovoice Console (https://picovoice.ai/console/)
```
Comment (laves) [inline demo/react-native/App.tsx:34]: ${YOUR_ACCESS_KEY_HERE}

File: demo/react-native/ios/PorcupineDemo/AppDelegate.m [hunk 9]
```diff
@@ -4,6 +4,10 @@
 #import <React/RCTBundleURLProvider.h>
 #import <React/RCTRootView.h>
 
+#if RCT_DEV
```
Comment (laves) [inline demo/react-native/ios/PorcupineDemo/AppDelegate.m:7]: what's this for??

**Review by laves — CHANGES_REQUESTED** [review]

File: demo/react-native/ios/PorcupineDemo/AppDelegate.m (hunk 9 above)
Comment (ksyeo1010) [reply demo/react-native/ios/PorcupineDemo/AppDelegate.m:7]: This was to remove a warning for dev environment when running iOS demo.

File: binding/react-native/ios/Porcupine.swift [hunk 10]
```diff
@@ -83,7 +96,50 @@ class PvPorcupine: NSObject {
...
+
+    private func getResourcePath(_ filePath: String) throws -> String {
+        if (!FileManager.default.fileExists(atPath: filePath)) {
+            if let resourcePath = Bundle.main.resourceURL?.appendingPathComponent("resources/\(filePath)").path {
+                if (FileManager.default.fileExists(atPath: resourcePath)) {
+                    return resourcePath
+                }
+            }
+            
+            throw PvPorcupineError.RuntimeError("Could not find file at path '\(filePath)'. If this is a packaged asset, ensure you have added it to your xcode project.")
```
Comment (laves) [inline binding/react-native/ios/Porcupine.swift:111]: It seems unnecessary to have that custom error for only this line. Maybe just throw a RuntimeError here and remove the custom error altogether

**Review by laves — CHANGES_REQUESTED** [review]

File: binding/react-native/ios/Porcupine.swift (hunk 10 above)
Comment (ksyeo1010) [reply binding/react-native/ios/Porcupine.swift:111]: I am not really sure there are 'default' error types in swift? I tried them as well search for it. don't think they exist.

File: README.md [hunk 11]
```diff
@@ -981,14 +981,14 @@ module to capture frames of audio and automatically pass it to the wake word eng
...
-`fromKeywords` and `fromKeywordPaths` static constructors.
+`fromBuiltInKeywords` and `fromKeywordPaths` static constructors.
 
 ```javascript
 const accessKey = "${ACCESS_KEY}" // AccessKey obtained from Picovoice Console (https://picovoice.ai/console/)
 
 async createPorcupine(){
     try{
-        this._porcupine = await Porcupine.fromKeywords(accessKey, ["picovoice"]);
+        this._porcupine = await Porcupine.fromBuiltInKeywords(accessKey, ["picovoice"]);
```
Comment (laves) [inline README.md:991]: update to BuiltInKeywords.PICOVOICE (and double-check the readmes for any more of these)

File: binding/react-native/src/porcupine.tsx [hunk 12]
```diff
@@ -41,7 +41,7 @@ class Porcupine {
    * [0, 1].
    * @returns An instance of the engine.
    */
-  public static async fromKeywords(
+  public static async fromBuildInKeywords(
```
Comment (laves) [inline binding/react-native/src/porcupine.tsx:44]: BuildIn -> BuiltIn

File: binding/react-native/src/porcupine.tsx [hunk 13]
```diff
@@ -50,11 +50,11 @@ class Porcupine {
     const keywordPaths = keywords.map((keyword) => {
       if (!Object.values(BuiltInKeywords).includes(keyword)) {
         throw new PorcupineExceptions.PorcupineInvalidArgumentException(
-          `keyword '${keyword}' is not a build-in keyword.`);
+          `keyword '${keyword}' is not a build-in keyword`);
```
Comment (laves) [inline binding/react-native/src/porcupine.tsx:53]: build-in -> built-in

File: binding/react-native/src/porcupinemanager.tsx [hunk 14]
```diff
@@ -38,15 +39,15 @@ class PorcupineManager {
    * [0, 1].
    * @returns An instance of the Porcupine Manager
    */
-  public static async fromKeywords(
+  public static async fromBuildInKeywords(
```
Comment (laves) [inline binding/react-native/src/porcupinemanager.tsx:42]: BuiltIn

**Review by laves — CHANGES_REQUESTED** [review]

**Review by laves — APPROVED** [review]

**Pull request merged**